proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,452 @@
+
+import BazComponent from './BazComponent.mjs';
+// Tries exact extension only
+```
+
+- As noted in *Subpath patterns*, the `"exports"` spec allows subpaths to use pattern trailers (`*`) strictly for substitution. `"exports"` values must also be file paths with extensions. Therefore packages cannot express subpaths where the import specifier maps to a file with an expanded extension.
+
+    ```js
+    "exports": {
+      "./subpath": "./subpath.js",
+      "./subpath.js": "./subpath.js",
+      "./subpath*": "./subpath*", // Insufficient to expand extension
+    }
+    ```
+
+- The second example (e.g. `.js.js`) may be seen as unintuitive, given the emergence of extensioned JS imports. We have the opportunity to omit this logic when resolving `"exports"`.
+
+**Proposed**:
+
+- **Breaking**: Under `"exports"`, we will remove support for platform-specific extensions.
+    - When resolving any import specifier:
+        - If the package defines `"exports"` and the exact specifier is matched, the package-defined path will be used.
+        - If there is no match in `"exports"`, Metro will look for files at the imported subpath, trying all extension variants (existing resolution logic).
+- Platform-specific extensions will narrow to a concept aimed at React Native app codebases, rather than for both apps and packages. We will communicate to React Native package authors that alternative patterns such as `Platform.select()` should be used.
+    - We have no near-term plans to drop platform-specific extensions for packages not using `"exports"`.
+- We will not take a strong stance on using extensionless or extensioned imports. The former may provide more flexibility for React Native package authors to change extensions in future without impacting consuming apps.
+
+#### Illustrated
+
+```json
+"exports": {
+  "./FooComponent": "./src/FooComponent.js",
+  "./FooComponent.js": "./src/FooComponent.js",
+}
+```
+
+Import specifiers listed in `"exports"` will be used when matched. Alternative paths will only be tried when there is no match in `"exports"`.
+
+```js
+import FooComponent from 'pkg/FooComponent';
+// Reads from "exports":
+//   pkg/src/FooComponent.js
+
+import FooComponent from 'pkg/FooComponent.js';
+// Reads from "exports":
+//   pkg/src/FooComponent.js
+
+import FooComponent from 'pkg/src/FooComponent';
+// No match in "exports" (Metro will print warning)
+// Tries files if present:
+//   pkg/src/FooComponent.[platform].js
+//   pkg/src/FooComponent.native.js
+//   pkg/src/FooComponent.js
+```
+
+The last example given is not expected to enable backwards compatibility, but may serendipitously capture pre-existing imports in apps. In a future strict mode, paths outside `"exports"` will not be considered.
+
+We will recommend that packages which rely on platform-specific extensions being available to consuming apps do not migrate to `"exports"` or change these entry points to handle platforms using `Platform.select()` or conditional exports.
```

Comment:
This seems to contradict the proposed implementation of conditional exports: _"We do not intend to use conditional exports as an alternative method to target the current React Native concept of platforms (e.g. `"android"`, `"ios"`, `"web"`)."_

Can we propose that people add a public "header" that re-exports the platform-forked files, e.g.

```js
// public.js
export * from "./platform-specific";

// platform-specific.android.js

// platform-specific.ios.js
```

Reply:
Partially — this refers mostly to the case of using the `"react-native"` condition to replace `.native.js` (and to a similar degree, `"browser"`). Our philosophy with this decision is to work well with packages in the wider npm ecosystem, and have slightly more opinion/restriction of patterns for React Native-specific packages.